### PR TITLE
Initial commit of Dart transformer to generate constructor stubs, 

### PR DESCRIPTION
**Pull request by kegluneq**
see https://github.com/angular/angular/issues/496.

This is just a starting point - it's still very rough. But we're making progress.

### CODE REVIEW

**Comment by mhevery** [comment]
Could you add tests, and some instructions as to how I could play with it? I would love to see what it generates.

**Comment by kegluneq** [comment]
html_transform_test.dart & transform_test.dart contain tests - were you looking for something in addition?

`testPhases` is kind of ugly - it takes
1) The test description
2) The transformer(s)
3) The assets to pass
4) The expected output.

File: modules/angular2/test/transform/transform_test.dart [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+  testPhases('Codegen for simple annotation', [[transform]], {
+    'a|web/index.html': htmlEntryPointContent,
+    'a|web/index.dart': '''
+        library web_foo;
+
+        import 'bar.dart';
+
+        void main() {
+          new Component2('Things');
```
Comment (mhevery) [inline modules/angular2/test/transform/transform_test.dart]: Where does `Component2` come from?
This should say: `bootstrap(Component);` 

`bootstrap` is what tells you where to start walking.

File: modules/angular2/test/transform/html_entry_point_test.dart [hunk 2]
```diff
@@ -0,0 +1,38 @@
+library angular2.test;
```
Comment (mhevery) [inline modules/angular2/test/transform/html_entry_point_test.dart]: Can you tell me about your plans to make these tests less dependent on large string comparisons (I expect that would be very easy to break the tests with just minor indentation changes, no?)

File: modules/angular2/test/transform/transform_test.dart [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        import 'index.dart' as i0;
+        import 'bar.dart' as i1;
+        import 'package:angular2/src/core/annotations/annotations.dart' as i2;
+        import 'foo.dart' as i3;
+
+        main() {
+          /* NOTE(tjblasi): Breaks function until we have "annotations" generated correctly.
+          reflector
+            ..registerType(i1.Component2, {
+              "factory": (i3.MyContext c) => new i1.Component2(c),
```
Comment (mhevery) [inline modules/angular2/test/transform/transform_test.dart]: So exciting! :-)

**Comment by mhevery** [comment]
Sorry, I somehow missed the tests the first time through. So to see the output I just do `pub build`?

**Comment by kegluneq** [comment]
Where does Component2 come from?
    This should say: bootstrap(Component);

    bootstrap is what tells you where to start walking.

    https://github.com/angular/angular/blob/master/modules/angular2/src/core/application.js#L101
Right, the traversal I'm using right now doesn't know about `bootstrap`, but I know I will need to address that. Component2 is hard-coded for the moment.

    Sorry, I somehow missed the tests the first time through. So to see the output I just do pub build?
Yes, pub build will show the generated code.

I've been using [pub serve](https://www.dartlang.org/tools/pub/cmd/pub-serve.html) & then hitting http://localhost:8080/src/hello_world/index.html

    Can you tell me about your plans to make these tests less dependent
    on large string comparisons (I expect that would be very easy to break
    the tests with just minor indentation changes, no?)
True, this is something I need to give a bit more thought to. Right before checking here I was updating both sides to use [package:dart_style](https://pub.dartlang.org/packages/dart_style), which feels a little hacky but I think is good enough for now.

**Comment by kegluneq** [comment]
By the way, if you want to see current output, check out the [annotate branch](https://github.com/kegluneq/angular/tree/annotate), which is where all my recent work is.

File: modules/angular2/transformer.dart [hunk 4]
```diff
@@ -0,0 +1,455 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
```
Comment (mhevery) [inline modules/angular2/transformer.dart]: can we move this into subfolder of `angular2` Maybe: `angular2/transformer/transformer.dart` (Or suggest something more appropriate.)

**Comment by kegluneq** [comment]
Moving to a subdir would be fine, however leaving it there in the Dart code would mean the difference between users declaring this in their pubspec:

<pre>
    - transformers:
        - angular2
</pre>

and this:

<pre>
    - transformers
        - angular2/src/transformer/transformer.dart
</pre>

What do you think about having `gulp` copy `transformer.dart` into `angular2/lib`?

**Pull request closed without merging**